Introduce condition variables

nashif opened the pull request:
Very similar to pthread_cond_wait with a mutex...

- [x] Docs
- [x] Samples
- [x] Tests
- [x] Usespace

Fixes #28406

nashif committed:
kernel: add condition variables

Introduce condition variables similar to how they are done in POSIX with
a mutex.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: kernel: add condition variable samples

Add a few samples for using condition variables..

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

pabigot commented on kernel/condvar.c:
```diff
@@ -0,0 +1,62 @@
+	k_spinlock_key_t key;
+
+	k_mutex_unlock(mutex);
+	key = k_spin_lock(&lock);
```
The standard semantics of condition variables requires that the unlock, queue, and wait be done atomically.  This code leaves a gap where the signal could be sent before the thread is queued to receive it, causing a failure.

You'd need to move the spin lock before the mutex to avoid that.

nashif replied:
yep, missed that.

andrewboie commented:
On a conceptual level this looks exactly like our futex API.

You have a wait API, and then two APIs: one that wakes one, and one that wakes all. Futex is the same, except the wake function has a boolean for wake one / wake all

andrewboie commented on include/kernel.h:
```diff
@@ -3714,6 +3714,32 @@ __syscall int k_mutex_unlock(struct k_mutex *mutex);
+
+
+#define K_CONDVAR_DEFINE(name) \
+	Z_STRUCT_SECTION_ITERABLE(k_condvar, name) = \
```
don't see any need for iteration at the moment, will this be added later?

nashif replied:
for object tracing eventually, but probably will drop it.

tyrantlizard77 reviewed (commented):
I think we might be able to make this and other IPC code even simpler if we had an internal abstraction for working with `_wait_q_t` objects. I had some old patches in this area I never submitted, let me see if I can dig them up

tbursztyka commented:
monitor objects could be added along with these condition variables (in another commit), at least it would make user's life simpler for instance on suspend/resume sample.

nashif committed:
tests: kernel: add test for condition variables

add a test for condition variables.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: add documentation for condition variables

Add documentation for condition variables in the synchronization section
of the kernel.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: add vrfy hooks to support userspace with condvar

Add needed vrfy hooks for userspace support.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

jukkar commented on include/kernel.h:
```diff
@@ -3439,6 +3439,83 @@ __syscall int k_mutex_unlock(struct k_mutex *mutex);
+__syscall int k_condvar_signal(struct k_condvar *condvar);
+
+/**
+ * @brief Unblock all threads all threads that are pending on the condition
```
typo, `all threads` is mentioned twice

jukkar commented on include/kernel.h:
```diff
@@ -3439,6 +3439,83 @@ __syscall int k_mutex_unlock(struct k_mutex *mutex);
+ * @brief Signals one thread that is pending on the condition variable
+ *
+ * @param condvar pointer to a @p k_condvar structure
+ * @retval 0 On success
```
Should we also document that all values !=0 are error, or just <0 is error? The documentation is ambiguous here

pabigot replied:
I like negative is error, non-negative is success, but agree that potential error returns should be documented.  As is the natural question is "if it can only return 0, why isn't it `void`?"

jukkar replied:
As a general note, I would argue that even if the function does not return <0, it might in the future if more error checks or conditions are added. In that case it would make sense to return int, even if we always return >= 0. It is difficult to change the API afterwards.

jukkar commented on include/kernel.h:
```diff
@@ -3439,6 +3439,83 @@ __syscall int k_mutex_unlock(struct k_mutex *mutex);
+ * @param mutex Address of the mutex.
+ * @param timeout Waiting period for the condition variable
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
+ * @return
```
Return value not documented

jukkar commented on kernel/condvar.c:
```diff
@@ -0,0 +1,72 @@
+
+	}
+	z_reschedule(&lock, key);
+	return woken;
```
What does the `woken` value mean here? According to documentation, this function returns 0 if ok, and any other value is error. According to this logic, if thread was woken, it is an error?

jukkar commented on kernel/condvar.c:
```diff
@@ -0,0 +1,72 @@
+
+	z_reschedule(&lock, key);
+
+	return woken;
```
Same issue here, the return value needs to be specified better in the documentation.

pabigot replied:
Yes: this is exactly why "non-negative is success" is so useful: to return the number of things affected.

jukkar replied:
The posix counterparts return only 0 for success and <0 for error. Do we really need to know the count here?

pabigot replied:
It's useful in testing, since there's no other indicator that something did or did not happen.  But if we're striving to match an existing API, maybe it's not right to do that.

jukkar replied:
If we have in the future `pthread_cond_xxx()` API which returns <0 on error and 0 on success, then it should be quite trivial to check that if `k_condvar_signal/broadcast()` returns >= 0, then return 0 in the pthread variant. So the zephyr API does not need to match exactly to pthread one in this case.

jukkar commented on samples/kernel/condition_variables/pm/src/main.c:
```diff
@@ -0,0 +1,163 @@
+
+	rc = release(my_id);
+	printk("Thread %ld returned after releasing. Status=%d\n", my_id, state);
+
```
extra empty line can be removed

pabigot commented on doc/reference/kernel/synchronization/condvar.rst:
```diff
@@ -0,0 +1,139 @@
+
+A condition variable is basically a queue of threads that threads can put
+themselves on when some state of execution (i.e., some condition) is not as
+desired (by waiting on the condition), the function
```
`s/, the function/. The function/` would help, but the parentheticals in first fragment still make it a bit hard to understand.

andyross commented on kernel/condvar.c:
```diff
@@ -0,0 +1,69 @@
+		arch_thread_return_value_set(thread, 0);
+		z_ready_thread(thread);
+	}
+	z_reschedule(&lock, key);
```
You can put the reschedule inside the test above, no need to do it if we didn't wake up a thread (though there have been consistent bike sheds regarding the way we document this).

pabigot reviewed (changes requested):
A few issues with documentation, otherwise this looks OK.

pabigot reviewed (changes requested):
I would like to see better documentation of the return values.  It looks a little like at least some of these could technically return `-EAGAIN` based on the behavior of `do_swap`, though I'm not sure.  If none of them can fail then some text like below (from  Linux `pthread_cond_wait` man page) would be worth making explicit:

>  pthread_cond_init,  pthread_cond_signal,  pthread_cond_broadcast,  and pthread_cond_wait never return an error code.

so we know not to insist that code check the result for errors.

I'd also like to see it based on #29668, but that's not a blocking desire.

pabigot commented on kernel/condvar.c:
```diff
@@ -0,0 +1,108 @@
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_condvar_init(struct k_condvar *condvar)
```
Coverage shows none of these are being invoked.

nashif replied:
<img width="636" alt="image" src="https://user-images.githubusercontent.com/180017/104985580-49764980-59df-11eb-8151-d93c5e2f1d6a.png">

pabigot replied:
Confirmed; possibly this was at a point were I didn't realize there were multiple tests so focused the -T argument too tightly.

pabigot commented on samples/kernel/condition_variables/condvar/src/main.c:
```diff
@@ -0,0 +1,105 @@
+		       __func__, my_id, count);
+		k_mutex_unlock(&count_mutex);
+
+		/* Do some work so threads can alternate on mutex lock */
```
s/Do some work/Sleep/ ?  Maybe this would be better as `k_yield()`.

pabigot commented on samples/kernel/condition_variables/pm/src/main.c:
```diff
@@ -0,0 +1,162 @@
+void suspend(long thread_id)
+{
+	printk("Suspending by thread %ld..\n", thread_id);
+	state = SUSPENDING;
```
I believe this is wrong as it's modifying a variable that elsewhere is protected by a mutex.  There's also general advice to do the broadcast/signal while the mutex is held to ensure consistency in scheduling.  https://stackoverflow.com/questions/4544234/ has some relevant information.

nashif replied:
this sample was intended to show wentong how this could be used for PM and I forgot about it. I removed that sample now.

pabigot commented on samples/kernel/condition_variables/pm/src/main.c:
```diff
@@ -0,0 +1,162 @@
+
+int claim(long thread_id)
+{
+
```
unnecessary blank line

pabigot reviewed (changes requested):
As in https://github.com/zephyrproject-rtos/zephyr/pull/27732#pullrequestreview-561290082 I'd still like to see an explanation why some of these functions are only documented to return 0.  Main blocker is the unlocked signalling in the pm module, which goes against the documented best practice.

nashif commented:
>  I'd still like to see an explanation why some of these functions are only documented to return 0.

future proofing of the API for additional error checking. Learned the hard way how difficult it was to change APIs to add error checking last time we went through this exercise.
